### PR TITLE
Add support for migration flag regarding task references

### PR DESCRIPTION
**Pull request by shawngustaw**
Uses ALM Connection endpoint and new backend code to determine
whether a sync has been performed with references. Once it has,
legacy lookup won't be performed.

Refs #SDE-3065

### CODE REVIEW

File: sdetools/alm_integration/alm_plugin_base.py [hunk 1]
```diff
@@ -782,6 +782,13 @@ def _check_supports_references(self, tasks):
 
         self.supports_references = False
 
+    def _check_has_migrated(self):
+        alm_connection = self.sde_plugin.get_alm_connection(self.alm_connection)
+        self.has_migrated = alm_connection.get('has_migrated', False)
```
Comment (sibartlett) [inline sdetools/alm_integration/alm_plugin_base.py]: If `alm_connection` has no `has_migrated` attribute, should we disable updating the migration status?

Then we can still run this code against v4.0?

File: sdetools/modules/sync_jira/jira_plugin.py [hunk 2]
```diff
@@ -166,7 +166,7 @@ def alm_add_task(self, task):
             new_issue['key'],
             reference=new_issue['id'],
             name=new_issue['key'],
-            link=self._get_issue_url(new_issue['key'])
+            link=new_issue['self']
```
Comment (shawngustaw) [inline sdetools/modules/sync_jira/jira_plugin.py]: @sibartlett / @geoffwhittington 

Do we want this link field to be "https://sdetest.atlassian.net/browse/SHAZ-13" or "https://sdetest.atlassian.net/rest/api/2/issue/50135"

Comment (geoffwhittington) [inline sdetools/modules/sync_jira/jira_plugin.py]: First one, we want the user-friendly link that a user can click and view the Jira issue in their browser.

File: sdetools/alm_integration/alm_plugin_base.py [hunk 3]
```diff
@@ -782,6 +782,13 @@ def _check_supports_references(self, tasks):
...
+
+    def _update_alm_migration_status(self, has_migrated):
+        self.sde_plugin.update_alm_connection(self.alm_connection, has_migrated)
```
Comment (MartinPetkov) [inline sdetools/alm_integration/alm_plugin_base.py]: It seems like ```sde_plugin.update_alm_connection``` and related functions can only update the has_migrated flag, so maybe change the name to something more specific? Perhaps ```update_alm_migration_status```, similar to this function.

File: sdetools/alm_integration/alm_plugin_base.py [hunk 4]
```diff
@@ -947,8 +957,13 @@ def update_progress(val, output=None):
 
             self._check_supports_references(tasks)
 
+            if self.supports_references:
+                self._check_has_migrated()
+
             logger.info('Filtered tasks')
 
+            self.created_references_during_sync = False
```
Comment (MartinPetkov) [inline sdetools/alm_integration/alm_plugin_base.py]: PyCharm suggests creating this in the __init__ method.

File: sdetools/sdelib/interactive_plugin.py [hunk 5]
```diff
@@ -228,6 +228,14 @@ def get_task_references(self, task_id):
         return self.api.get_task_references(self.prj_id, task_id)['results']
 
     @_verify_connect
+    def get_alm_connection(self, connection_id):
+        return self.api.get_alm_connection(connection_id)
+
+    @_verify_connect
+    def update_alm_connection(self, connection_id, has_migrated):
```
Comment (MartinPetkov) [inline sdetools/sdelib/interactive_plugin.py]: Same comment as above about the method name.

Comment (geoffwhittington) [inline sdetools/sdelib/interactive_plugin.py]: Just FYI, we'll be working on this area of functionality again very soon.

**Comment by geoffwhittington** [comment]
Can you squash before merge?

**Comment by shawngustaw** [comment]
@geoffwhittington yupp. Just working on fixing TravisCI before that.

**Comment by geoffwhittington** [comment]
This is going to be awesome.

**Comment by geoffwhittington** [comment]
What's the status of this PR, are we merging this into 4.1 or 4.2?

**Comment by shawngustaw** [comment]
@geoffwhittington I just haven't been able to get the tests pass mostly because I'm not too familiar with how they're structured. Going to try to get this wrapped up today though.

**Pull request merged**